shards/shards.go

```diff
@@ -667,13 +667,71 @@ search:
 			r.Priority = r.priority
 			r.MaxPendingPriority = pending.max()
 
-			sender.Send(r.SearchResult)
+			// For simple shards, send the results immediately.
+			if len(r.SearchResult.RepoURLs) <= 1 {
+				sender.Send(r.SearchResult)
+				break
+			}
+
+			// For compound shards, stream results per repo.
+			stats, srs := splitByRepository(r.SearchResult)
+			tr.LazyPrintf("compound shard: results from %d repositories", len(srs))
+			for _, sr := range srs {
+				if len(sr.Files) == 0 {
+					panic("unreachable")
+				}
+				sr.Priority = sr.Files[0].RepositoryPriority
+
+				// TODO (stefan): make sure repos in compound shards are searched in order of their priority.
+				sr.MaxPendingPriority = r.MaxPendingPriority
+				sender.Send(sr)
+			}
+			sender.Send(&zoekt.SearchResult{Stats: stats})
 		}
 	}
 
 	return func() { runtime.KeepAlive(shards) }, err
 }
 
+// splitByRepository splits a zoekt.SearchResult by repository ID. We separate
+// the stats from the searchResult, because stats must be aggregateable.
+func splitByRepository(result *zoekt.SearchResult) (zoekt.Stats, []*zoekt.SearchResult) {
+	if len(result.Files) == 0 {
+		return result.Stats, nil
+	}
+	if len(result.Files) == 1 {
+		stats := result.Stats
+		result.Stats = zoekt.Stats{}
+		return stats, []*zoekt.SearchResult{result}
+	}
+
+	res := []*zoekt.SearchResult{}
+	appendResult := func(repoName string, a, b int) {
+		res = append(res, &zoekt.SearchResult{
+			Files:         result.Files[a:b],
+			RepoURLs:      map[string]string{repoName: result.RepoURLs[repoName]},
+			LineFragments: map[string]string{repoName: result.LineFragments[repoName]},
+		})
+	}
+
+	var startIndex, endIndex int
+	curRepoID := result.Files[0].RepositoryID
+	curRepoName := result.Files[0].Repository
+
+	fm := zoekt.FileMatch{}
+	for endIndex, fm = range result.Files {
```

Comment:
out of interest, how does the ranking by `Score` affect this? I can't remember, but we may be sorting within a shard? This means after sorting you might not have all Files from a single repo being contiguous. However, I am likely confused and we don't do ranking in indexData :)

Reply:
Yep that's true. `Score` breaks this. Good find! Either we move sorting out of indexData.Search to here or we move streaming to indexData.

Reply:
Update: as discussed in our sync meeting, I moved sorting from indexedSearch one level up to shardedSearcher.